core/src/main/java/org/jdbi/v3/core/async/AbstractJdbiExecutor.java

```diff
@@ -0,0 +1,93 @@
+ * limitations under the License.
+ */
+package org.jdbi.v3.core.async;
+
+import java.util.concurrent.CompletionStage;
+
+import org.jdbi.v3.core.HandleCallback;
+import org.jdbi.v3.core.HandleConsumer;
+import org.jdbi.v3.core.Jdbi;
+import org.jdbi.v3.core.extension.ExtensionCallback;
+import org.jdbi.v3.core.extension.ExtensionConsumer;
+import org.jdbi.v3.core.transaction.TransactionIsolationLevel;
+
+public abstract class AbstractJdbiExecutor implements JdbiExecutor {
+
+    /**
+     * Single method through which all other methods converge. If you want to override AbstractJdbiExecutor, you only need to override this one method.
+     *
+     * @param handler the handler that takes a Jdbi instance and returns a value
+     * @param <R>     type returned by the callback
+     * @param <X>     exception type thrown by the callback, if any.
+     * @return a completion stage that will complete when the handler returns or throws an exception
+     */
+    protected abstract <R, X extends Exception> CompletionStage<R> doExecute(Handler<R, X> handler);
+
+    @Override
+    public <R, X extends Exception> CompletionStage<R> withHandle(final HandleCallback<R, X> callback) {
+        return doExecute(jdbi -> jdbi.withHandle(callback));
+    }
+
+    @Override
+    public <R, X extends Exception> CompletionStage<R> inTransaction(final HandleCallback<R, X> callback) {
+        return doExecute(jdbi -> jdbi.inTransaction(callback));
+    }
+
+    @Override
+    public <R, X extends Exception> CompletionStage<R> inTransaction(final TransactionIsolationLevel level, final HandleCallback<R, X> callback) {
+        return doExecute(jdbi -> jdbi.inTransaction(level, callback));
+    }
+
+    @Override
+    public <X extends Exception> CompletionStage<Void> useHandle(final HandleConsumer<X> consumer) {
+        return doExecute(jdbi -> {
+            jdbi.useHandle(consumer);
+            return null;
+        });
+    }
+
+    @Override
+    public <X extends Exception> CompletionStage<Void> useTransaction(final HandleConsumer<X> callback) {
+        return doExecute(jdbi -> {
+            jdbi.useTransaction(callback);
+            return null;
+        });
+    }
+
+    @Override
+    public <X extends Exception> CompletionStage<Void> useTransaction(final TransactionIsolationLevel level, final HandleConsumer<X> callback) {
+        return doExecute(jdbi -> {
+            jdbi.useTransaction(level, callback);
+            return null;
+        });
+    }
+
+    @Override
+    public <R, E, X extends Exception> CompletionStage<R> withExtension(final Class<E> extensionType, final ExtensionCallback<R, E, X> callback) {
+        return doExecute(jdbi -> jdbi.withExtension(extensionType, callback));
+    }
+
+    @Override
+    public <E, X extends Exception> CompletionStage<Void> useExtension(final Class<E> extensionType, final ExtensionConsumer<E, X> callback) {
+        return doExecute(jdbi -> {
```

Comment:
for most of our code, we use two distinct interfaces (one would be the `JdbiCallback` below, the other would be

```
public interface JdbiConsumer<X extends Exception> {
    void useJdbi(Jdbi jdbi) throws X;
}
```

The second one should be used here, so this would fold into 

```
return doExecute(jdbi -> jdbi.useExtension(extensionType, callback));
```

Reply:
Changed to use withExecute and useExecute